### PR TITLE
Testing hpo packages

### PR DESCRIPTION
**Pull request by allaffa**
This PR introduces three main examples for hyperparameter optimization (HPO) using the QM9 dataset:

1. `qm9_optuna.py`: HPO example using `optuna`
2. `qm9_deephyper.py`: HPO example running in sequential using `deephyper`
3. `qm9_deephyper_multi.py`: HPO example running in parallel using `deephyper`

### CODE REVIEW

**Comment by jychoi-hpc** [comment]
Can we save this qm9 as a separate directory (say, qm9_hpo)? We  often use qm9 as a minimum case. It would be great if we make "qm9" as simple as possible. How do you think?

**Comment by allaffa** [comment]
> Can we save this qm9 as a separate directory (say, qm9_hpo)? We often use qm9 as a minimum case. It would be great if we make "qm9" as simple as possible. How do you think?

@jychoi-hpc agree.

**Review by jychoi-hpc — APPROVED** [review]
Thank you for the update. It looks good to me.

**Review by jychoi-hpc — APPROVED** [review]
It looks good to me. Thanks.

File: examples/qm9_hpo/qm9_deephyper_multi.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    # NOTE: data is moved to the device in the pre-transform.
+    # NOTE: transforms/filters will NOT be re-run unless the qm9/processed/ directory is removed.
+    dataset = torch_geometric.datasets.QM9(
+        root="dataset/qm9", pre_transform=qm9_pre_transform
+    )
+
+    trainset, valset, testset = hydragnn.preprocess.split_dataset(dataset, 0.8, False)
+    (train_loader, val_loader, test_loader) = hydragnn.preprocess.create_dataloaders(
+        trainset, valset, testset, 64
+    )
```
Comment (pzhanggit) [inline examples/qm9_hpo/qm9_deephyper_multi.py:134]: Could we pass the data loaders defined here as command line arguments in `qm9.py` if we want each trail to reuse the same loaders?

File: examples/qm9_hpo/qm9.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    optimizer,
+    train_loader,
+    val_loader,
+    test_loader,
+    writer,
+    scheduler,
+    config["NeuralNetwork"],
+    log_name,
+    verbosity,
+)
```
Comment (pzhanggit) [inline examples/qm9_hpo/qm9.py:114]: `save_model` after training

**Review by pzhanggit — COMMENTED** [review]
The PR looks good to me overall. I'll need to learn more about these hpo tools. Three general comments: 1) remove or clean up unused functions in `dh_utils.py`; 2) save trained models in all trials in `qm9.py`; 3) maybe add data loaders as command line arguments to avoid repeated data preprocessing in trails.

File: examples/qm9_hpo/qm9_deephyper_multi.py (hunk 1 above)
Comment (allaffa) [reply examples/qm9_hpo/qm9_deephyper_multi.py:134]: For large scale runs, where we data loaders are split across processes, I do not see a practical way to do this.

File: examples/qm9_hpo/qm9.py (hunk 2 above)
Comment (allaffa) [reply examples/qm9_hpo/qm9.py:124]: Thanks. I added these lines.

**Comment by allaffa** [comment]
@pzhanggit 
I also added num_headlayers and dim_headlayers as hyper-parameters

**Review by jychoi-hpc — APPROVED** [review]
Great. It looks good to me.

**Pull request merged**